src/arch/wasm/CodeGen.zig

```diff
@@ -4055,189 +4010,218 @@ fn airStructFieldVal(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
-        body: []const Air.Inst.Index,
-    }).initCapacity(func.gpa, switch_br.cases_len);
-    defer for (case_list.items) |case| {
-        func.gpa.free(case.values);
-    } else case_list.deinit();
-
-    var lowest_maybe: ?i32 = null;
-    var highest_maybe: ?i32 = null;
+    const has_else_body = switch_br.else_body_len != 0;
+    const branch_count = switch_br.cases_len + 1; // if else branch is missing, we trap when failing all conditions
+    try func.branches.ensureUnusedCapacity(func.gpa, switch_br.cases_len + @intFromBool(has_else_body));
+
+    if (switch_br.cases_len == 0) {
+        assert(has_else_body);
+
+        var it = switch_br.iterateCases();
+        const else_body = it.elseBody();
+
+        func.branches.appendAssumeCapacity(.{});
+        const else_deaths = liveness.deaths.len - 1;
+        try func.currentBranch().values.ensureUnusedCapacity(func.gpa, liveness.deaths[else_deaths].len);
+        defer {
+            var else_branch = func.branches.pop();
+            else_branch.deinit(func.gpa);
+        }
+        try func.genBody(else_body);
+
+        if (is_dispatch_loop) {
+            try func.endBlock(); // dispatch loop end
+        }
+        return func.finishAir(inst, .none, &.{});
+    }
+
+    const Case = Air.UnwrappedSwitch.CaseIterator.Case;
+    const cases = try func.gpa.alloc(Case, switch_br.cases_len);
+    defer func.gpa.free(cases);
+
+    const big = std.math.big;
+    const BigIntManaged = big.int.Managed;
+    const BigIntConst = big.int.Const;
+    const BigIntSpace = Value.BigIntSpace;
+    const limb_count = big.int.calcTwosCompLimbCount(@intCast(target_ty.bitSize(zcu)));
+
+    var lowest_tmp_bigint = try BigIntManaged.initCapacity(func.gpa, limb_count + 1);
+    defer lowest_tmp_bigint.deinit();
+    var highest_tmp_bigint = try BigIntManaged.initCapacity(func.gpa, limb_count + 1);
+    defer highest_tmp_bigint.deinit();
+    var tmp_bigint = try BigIntManaged.initCapacity(func.gpa, limb_count + 1);
+    defer tmp_bigint.deinit();
+    var tmp_space_0: BigIntSpace = undefined;
+    var tmp_space_1: BigIntSpace = undefined;
+
+    var is_first_value = true;
+    var lowest_value_ref: Air.Inst.Ref = undefined;
+    var lowest: BigIntConst = undefined;
+    var highest: BigIntConst = undefined;
+    var branching_size: u32 = 0; // single item +1, range +2
+
     var it = switch_br.iterateCases();
     while (it.next()) |case| {
-        const values = try func.gpa.alloc(CaseValue, case.items.len + case.ranges.len);
-        errdefer func.gpa.free(values);
-
-        for (case.items, 0..) |ref, i| {
-            const item_val = (try func.air.value(ref, pt)).?;
-            const int_val = func.valueAsI32(item_val);
-            if (lowest_maybe == null or int_val < lowest_maybe.?) {
-                lowest_maybe = int_val;
+        cases[case.idx] = case;
+
+        for (case.items) |ref| {
+            const value_bigint = (try func.air.value(ref, pt)).?.toBigInt(&tmp_space_0, zcu);
```

Comment:
This is going to be too inefficient. If the value isn't going to fit in the range of a `u32`, we should simply stop attempting to lower all the values and abort attempting to use a `br_table`. I think we can come up with a way better design than relying on big-integers.

Reply:
Are you proposing to not do `br_table` for switches that branch on > 32 bits values? I think it's okay to detect switches with high values, but that are densely located, just subtracting the lowest value will make it work for .br_table. I think we should do such optimization for u64/i64/u128/i128.

Reply:
Regarding big-integers inefficiency, we are still way below from `Sema.zig`, because it will do even more checks on intersections for each switch case.

Reply:
I don't think being faster than Sema is a good enough argument. However, considering the new implementation does improve status quo by a lot, and you did your due diligence, we can go with the bigint strategy. I do think it's worth investing time to determine whether it's worth it to have a non-bigint implementation for the fast path. The majority of user-land code involving switches seems to be on enum values <32bit and it might be worth optimising for that use case.